h/services/group.py

```diff
@@ -88,21 +89,31 @@ def groupids_readable_by(self, user, group_ids=None):
 
         :type user: `h.models.user.User`
         """
-        readable = Group.readable_by == ReadableBy.world
+        # Very few groups are readable by world, query those separately
+        readable_by_world_query = select(Group.pubid).where(
+            Group.readable_by == ReadableBy.world
+        )
+        query = readable_by_world_query
 
         if user is not None:
-            readable_member = sa.and_(
-                Group.readable_by == ReadableBy.members,
-                Group.members.any(User.id == user.id),
+            user_is_member_query = (
+                select(Group.pubid)
+                .join(GroupMembership)
+                .where(
+                    GroupMembership.user_id == user.id,
+                    Group.readable_by == ReadableBy.members,
+                )
+            )
+            # Union these with the world readable ones
+            # We wrap a subquery around in case we need to apply the group_ids filter below
+            query = select(
+                union(readable_by_world_query, user_is_member_query).subquery()
             )
-            readable = sa.or_(readable, readable_member)
 
         if group_ids:
-            readable = sa.and_(Group.pubid.in_(group_ids), readable)
+            query = query.where(literal_column("pubid").in_(group_ids))
```

Comment:
This literal_column business allows to filter by both the query with user and the one without it.